codegen/codegen.go

```diff
@@ -300,15 +300,18 @@ func emitFilesystemSourceStmt(info *CodeGenInfo, scope *ast.Scope, call *ast.Cal
 			return st, err
 		}
 
-		id := identity.NewID()
-		info.Locals[id] = path
-
 		var opts []llb.LocalOption
 		for _, iopt := range iopts {
 			opt := iopt.(llb.LocalOption)
 			opts = append(opts, opt)
 		}
 
+		// get a consistent hash for this local (path + options) so we don't
+		// transport the same content multiple times when referenced repeatedly
+		sum := sha256.Sum256([]byte(fmt.Sprintf("%s:%#v", path, opts)))
```

Comment:
I will keep working on this, the salt string ends up looking like: `.:[]llb.LocalOption{(llb.localOptionFunc)(0x1a2cae0)}` which does not seem like a great value to hash.

Reply:
Yeah, you could try:
```go
st = llb.Image(id, opts...)

dgst, _, _, err := st.Output().Vertex().Marshal(&llb.Constraints{})
if err != nil {
    return st, err
}

return st, nil
```

Reply:
Hm.. yeah, kind of a chicken and egg problem here. The `llb.Local` gives a `llb.State` that we can get a digest, but we also need it to create the `llb.Local`.